src/audio/pipewire.c

```diff
@@ -0,0 +1,322 @@
+            break;
+        case SPD_AUDIO_BE:
+            format = SPA_AUDIO_FORMAT_S16_BE;
+            break;
+        default:
+            pw_log_error("invalid audio format specifier");
+            return -1;
+        }
+    }
+    else if (track.bits == 8)
+    {
+        state->stride = track.num_channels * sizeof(uint8_t);
+        format = SPA_AUDIO_FORMAT_S8;
+    }
+    else
+    {
+        pw_log_error("invalid audio format specifier");
+        return -1;
+    }
+
+    params = spa_format_audio_raw_build(&b, SPA_PARAM_EnumFormat, &SPA_AUDIO_INFO_RAW_INIT(.format = format, .channels = track.num_channels, .rate = track.sample_rate));
+    pw_thread_loop_lock(state->loop);
+    pw_stream_connect(state->stream, PW_DIRECTION_OUTPUT, PW_ID_ANY, PW_STREAM_FLAG_AUTOCONNECT | PW_STREAM_FLAG_MAP_BUFFERS | PW_STREAM_FLAG_RT_PROCESS, &params, 1);
+    pw_thread_loop_unlock(state->loop);
+    return 0;
+}
+static int pipewire_feed_sink_overlap(AudioID *id, AudioTrack track)
+{
+    module_state *state = (module_state *)id;
+    uint32_t write_index, fill_quantity, room_left_in_buffer = 0, track_buffer_size;
+    uint64_t dummy; // used to read from the event file descripter. We don't use this for anything, at the moment, it's always 42
+    // if the stream has been deactivated because of a pipewire_stop call, we activate it
+    pw_thread_loop_lock(state->loop);
+    if (pw_stream_get_state(state->stream, NULL) == PW_STREAM_STATE_PAUSED)
+    {
+        pw_stream_set_active(state->stream, true);
+    }
+    pw_thread_loop_unlock(state->loop);
+    track_buffer_size = track.num_samples;
+    // as long as we have samples in the buffer sent by speech dispatcher, we loop, blocking spd from calling us till the audio system played what we loaded in it
+    while (track_buffer_size > 0)
+    {
+        // drainage spinlock!
+        // we wait till fill_quantity is not empty after this thread can continue, aka after on_process updated that ringbuffer index with what it managed to play, so that we have some room in the buffer to fill from what's remaining of what was given to us by speech dispatcher
+        while (true)
+        {
+            fill_quantity = spa_ringbuffer_get_write_index(&state->rb, &write_index);
+            // see if there's some room in the buffer, if so, stop spinning
+            room_left_in_buffer = SAMPLE_BUFFER_SIZE - fill_quantity;
+            if (room_left_in_buffer > 0)
+                // then we have room in the buffer, so stop spinning
+                break;
+            // otherwise, to not starve a cpu core of resources, we suspend our spinning until something, in this case on_process, wakes us up. In an embedded system without these primitives, spinning endlessly like that would be the only option. We could do the same here, but for efficiency reasons alone, we don't
+            pw_thread_loop_lock(state->loop);
+            struct pw_loop *program_loop = pw_thread_loop_get_loop(state->loop);
```

Comment:
you could record the `pw_loop` in `state` too, to avoid having to lock each time

Reply:
That'd be really meaningless here: you're releasing the lock before using `program_loop`. If `program_loop` was ever to change, you'd be screwed right after you unlock. Pipewire allocates the program loop on `pw_thread_loop_new`, and that never changes after that.

Reply:
alright, will do then